### PR TITLE
Specify how the compiler identifies the core library.

### PR DESCRIPTION
**Pull request by gafter**
@AlekseyTs @tmat Please review
/cc @dotnet/roslyn-compiler

### CODE REVIEW

**Comment by gafter** [comment]
Fixes #13279

File: docs/compilers/Co-located core types.md [hunk 1]
```diff
@@ -15,7 +15,7 @@ Scenarios
 
 ## Design
 
-Compilers may assume that there is exactly one assembly representing the "core assembly". This core assembly is either referenced or is currently being compiled.
+Compilers may assume that there is exactly one assembly representing the "core assembly". This core assembly is either referenced or is currently being compiled. The unique core assembly may be identified by the compiler as the first referenced assembly that does not reference any other assembly, or the assembly being compiled if there are no referenced assemblies.
```
Comment (tmat) [inline docs/compilers/Co-located core types.md]: Don't we check that we have a single cor lib (as opposed to picking the first one)?

Perhaps we could be more rigorous here? How about we define what core assembly means and what the requirements are on it? Such definition should imo say 

> *core assembly* is an assembly that defines System.Object
> core assembly is required to have no references

Comment (gafter) [inline docs/compilers/Co-located core types.md]: The precise rules that we use is that we look for an assembly that satisfies the following constraints, and require there be a unique candidate ([CommonReferenceManager.Binding.cs line 897](https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Binding.cs#L897)):
- May not be a linked reference
- Must not reference any other assemblies
- Must not contain any NoPia types
- Some other condition about not being superseded
- Defines System.Object

The native compiler allowed more than one candidate and selected the first one, but we never reproduced that behavior in Roslyn.

**Comment by gafter** [comment]
@tmat My intent is not to document precisely what the compiler does, but to document the requirements that we can place on the platform. Specifically, I was aiming to document the requirement that corlib may not reference other assemblies.

Do you think this is good as written? If not, do you have specific suggestions for change?

**Comment by tmat** [comment]
Then let's not mention the last thing, that each compilation needs to have one corlib. I see now your point though and see the requirement in the statement. However, it wasn't quite clear to me since the focus of the sentence is on the compiler, not on the corlib. I'd rather avoid mentioning the compiler as the fact whether an assembly is a corlib or not doesn't depend on what tool is looking at it.

Something as simple as: "Core assembly is an assembly that defines System.Object. Core assembly is required to have no references." would imo be better.

**Comment by gafter** [comment]
> Something as simple as: "Core assembly is an assembly that defines System.Object. Core assembly is required to have no references." would imo be better.

The doc already says "The core assembly contains the type definitions for the subset of built-in and special types supported on a platform."

So I'll just add the sentence "The core assembly may not reference any other assemblies."

**Comment by gafter** [comment]
@AlekseyTs @tmat Please let me know if this tiny revision of the docs looks good.

**Comment by tmat** [comment]
:+1:

**Comment by AlekseyTs** [comment]
LGTM

**Pull request merged**